src/python/pants/engine/isolated_process.py

```diff
@@ -23,28 +24,25 @@ class ExecuteProcessRequest(datatype([
   ('argv', tuple),
   ('input_files', DirectoryDigest),
   ('description', text_type),
-  ('env', tuple),
-  ('output_files', tuple),
-  ('output_directories', tuple),
+  # TODO: allow inferring a default value if a type is provided which has a 0-arg constructor.
+  F('env', Exactly(tuple, dict, type(None)), default_value=None),
+  F('output_files', tuple, default_value=()),
+  F('output_directories', tuple, default_value=()),
   # NB: timeout_seconds covers the whole remote operation including queuing and setup.
-  ('timeout_seconds', Exactly(float, int)),
-  ('jdk_home', Exactly(text_type, type(None))),
+  F('timeout_seconds', Exactly(float, int), default_value=_default_timeout_seconds),
+  F('jdk_home', Exactly(text_type, type(None)), default_value=None),
 ])):
   """Request for execution with args and snapshots to extract."""
 
-  def __new__(
-    cls,
-    argv,
-    input_files,
-    description,
-    env=None,
-    output_files=(),
-    output_directories=(),
-    timeout_seconds=_default_timeout_seconds,
-    jdk_home=None,
-  ):
+  def __new__(cls, *args, **kwargs):
```

Comment:
I find this this implementation to be more work for me to read, rather than less... Before it massaged param, and then constructed an object as I'd expect; now it has multiple layers of validation, and the provision on the type definition for an intermediate state which would be invalid outside the constructor (having a dict as a field value)...

Maybe the answer here is to fall back to the automatically implemented `__new__`, and have a `with_env` class-function which does the env munging and passes the tuple to the constructor?